hub/saml-engine/src/integration-test/java/uk/gov/ida/integrationtest/hub/samlengine/apprule/IdpAuthnResponseTranslatorResourceWithRedisTest.java

```diff
@@ -0,0 +1,176 @@
+
+        clientResponse = postToSamlEngine(samlResponseDto);
+
+        assertThat(clientResponse.getStatus()).isEqualTo(Response.Status.BAD_REQUEST.getStatusCode());
+        ErrorStatusDto errorStatusDto = clientResponse.readEntity(ErrorStatusDto.class);
+        assertThat(errorStatusDto.getExceptionType()).isEqualTo(ExceptionType.INVALID_SAML);
+    }
+
+    @Test
+    public void handleResponseFromIdp_shouldThrowExceptionIfAuthnStatementAssertionIsReplayedInResponseFromIdp() throws Exception {
+        String authnStatementAssertionId = "authnStatementAssertionId" + UUID.randomUUID().toString();
+        String mdsStatementAssertionId = "mdsStatementAssertionId" + UUID.randomUUID().toString();
+        SamlAuthnResponseTranslatorDto samlResponseDto_1 = getSuccessSamlAuthnResponseTranslatorDto(STUB_IDP_ONE, authnStatementAssertionId, mdsStatementAssertionId + "-1");
+        SamlAuthnResponseTranslatorDto samlResponseDto_2 = getSuccessSamlAuthnResponseTranslatorDto(STUB_IDP_ONE, authnStatementAssertionId, mdsStatementAssertionId + "-2");
+
+        Response clientResponse = postToSamlEngine(samlResponseDto_1);
+
+        assertThat(clientResponse.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
+
+        clientResponse = postToSamlEngine(samlResponseDto_2);
+
+        assertThat(clientResponse.getStatus()).isEqualTo(Response.Status.BAD_REQUEST.getStatusCode());
+        ErrorStatusDto errorStatusDto = clientResponse.readEntity(ErrorStatusDto.class);
+        assertThat(errorStatusDto.getExceptionType()).isEqualTo(ExceptionType.INVALID_SAML);
+    }
+
+    @Test
+    public void handleResponseFromIdp_shouldThrowExceptionIfmdsAssertionIsReplayedInResponseFromIdp() throws Exception {
+        String authnStatementAssertionId = "authnStatementAssertionId" + UUID.randomUUID().toString();
+        String mdsStatementAssertionId = "mdsStatementAssertionId" + UUID.randomUUID().toString();
+        SamlAuthnResponseTranslatorDto samlResponseDto_1 = getSuccessSamlAuthnResponseTranslatorDto(STUB_IDP_ONE, authnStatementAssertionId + "-1", mdsStatementAssertionId);
+        SamlAuthnResponseTranslatorDto samlResponseDto_2 = getSuccessSamlAuthnResponseTranslatorDto(STUB_IDP_ONE, authnStatementAssertionId + "-2", mdsStatementAssertionId);
+
+        Response clientResponse = postToSamlEngine(samlResponseDto_1);
+
+        assertThat(clientResponse.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
+
+        clientResponse = postToSamlEngine(samlResponseDto_2);
+
+        assertThat(clientResponse.getStatus()).isEqualTo(Response.Status.BAD_REQUEST.getStatusCode());
+        ErrorStatusDto errorStatusDto = clientResponse.readEntity(ErrorStatusDto.class);
+        assertThat(errorStatusDto.getExceptionType()).isEqualTo(ExceptionType.INVALID_SAML);
+    }
+
+    @Test
+    public void handleResponseFromIdp_shouldThrowExceptionForSecondIdpIfTwoIdpsSubmitAnAuthnStatementAssertionWithTheSameId() throws Exception {
```

Comment:
Do we have a test case for 2 different IDPs submitting MDS Statement Assertions sharing the same id?